### PR TITLE
Add support to produce WebAssembly runtime assets

### PR DESCRIPTION
**Pull request by safern**
This enables a build for webassembly as a OSGroup-ArchGroup using a Linux implementation detail from our sources and produces a package with System.Native pal as a native asset using `emscripten` to build it.

The package would be `runtime.webassembly-wasm.Microsoft.Private.CoreFx.NETCoreApp`.

Worked in collaboration with @vargaz in order to tweak the native sources to have a clean build.

cc: @danmosemsft @marek-safar @lewing

### CODE REVIEW

File: Directory.Build.props [hunk 1]
```diff
@@ -67,8 +67,12 @@
     <HostArch>$([System.Runtime.InteropServices.RuntimeInformation]::ProcessArchitecture.ToString().ToLowerInvariant)</HostArch>
     <ArchGroup Condition="'$(ArchGroup)' == '' AND '$(HostArch)' == 'arm'">arm</ArchGroup>
     <ArchGroup Condition="'$(ArchGroup)' == '' AND '$(HostArch)' == 'arm64'">arm64</ArchGroup>
+    <ArchGroup Condition="'$(ArchGroup)' == '' AND '$(OSGroup)' == 'WebAssembly'">wasm</ArchGroup>
     <ArchGroup Condition="'$(ArchGroup)' == ''">x64</ArchGroup>
 
+    <!-- Runtime OS is calculated based on the build system OS, however if building for WebAssembly we need to let the build system that. -->
```
Comment (stephentoub) [inline Directory.Build.props]: Nit: "we need to let the build system that" is missing a word

File: src/Native/Unix/System.Native/pal_networking.c [hunk 2]
```diff
@@ -2492,8 +2492,8 @@ static int32_t WaitForSocketEventsInner(int32_t port, SocketEvent* buffer, int32
 }
 
 #else
-#warning epoll/kqueue not detected; building with stub socket events support
-static const size_t SocketEventBufferElementSize = sizeof(struct pollfd);
+//#warning epoll/kqueue not detected; building with stub socket events support
```
Comment (stephentoub) [inline src/Native/Unix/System.Native/pal_networking.c]: Did you mean to leave this commented out?

**Review by stephentoub — COMMENTED** [review]
Thanks, @safern.

File: Directory.Build.props [hunk 3]
```diff
@@ -140,8 +144,9 @@
     <!-- We build linux-musl-arm on a ubuntu container, so we can't use the toolset build for alpine runtime. We need to use portable linux RID for our toolset in order to be able to use it. -->
     <ToolRuntimeRID Condition="'$(_runtimeOS)' == 'linux-musl' AND $(ArchGroup.StartsWith('arm')) AND !$(HostArch.StartsWith('arm'))">linux-x64</ToolRuntimeRID>
 
-    <!-- There are no WebAssembly tools, so treat them as Windows -->
-    <ToolRuntimeRID Condition="'$(RuntimeOS)' == 'WebAssembly'">win-x64</ToolRuntimeRID>
+    <!-- There are no WebAssembly tools, so use the default ones -->
+    <ToolRuntimeRID Condition="'$(RuntimeOS)' == 'webassembly' AND '$(OS)' == 'Windows_NT'">win-x64</ToolRuntimeRID>
+    <ToolRuntimeRID Condition="'$(RuntimeOS)' == 'webassembly' AND '$(OS)' != 'Windows_NT'">linux-x64</ToolRuntimeRID>
```
Comment (ericstj) [inline Directory.Build.props]: What about mac?

File: eng/restore/runtime/runtime.depproj [hunk 4]
```diff
@@ -1,6 +1,7 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
     <RuntimeIdentifier>$(PackageRID)</RuntimeIdentifier>
+    <RuntimeIdentifier Condition="'$(RuntimeOS)' == 'webassembly'">$(ToolRuntimeRID)</RuntimeIdentifier>
```
Comment (ericstj) [inline eng/restore/runtime/runtime.depproj]: This deserves a comment.  We're using ToolRuntime as a placeholder for the real corelib/runtime components until we have an actual set of runtime bits to consume for webassembly.

**Comment by ericstj** [comment]
WebAssembly derives from Unix: https://github.com/dotnet/corefx/blob/c98870374325bb3e55cf764757826ba574d7186a/eng/configurations/osgroups.props#L37-L41

We may need to add IL configurations to projects which are cross-compiling by Win/OSX/Linux since they'll not have a RID specific implementation for WebAssembly.

File: Directory.Build.props (hunk 3 above)
Comment (safern) [reply Directory.Build.props]: I'll add support for building wasm on mac.

File: eng/restore/runtime/runtime.depproj (hunk 4 above)
Comment (safern) [reply eng/restore/runtime/runtime.depproj]: Makes sense.

Comment (safern) [reply eng/restore/runtime/runtime.depproj]: Makes sense.

File: src/Native/Unix/System.Native/pal_networking.c (hunk 2 above)
Comment (safern) [reply src/Native/Unix/System.Native/pal_networking.c]: Nope. I'll remove. Thanks

File: Directory.Build.props (hunk 1 above)
Comment (safern) [reply Directory.Build.props]: Thanks, will update.

**Comment by safern** [comment]
> WebAssembly derives from Unix:

I'm changing it to derive from Linux and set `TargetsLinux=true` on this PR. I think that would be sufficient, no?

https://github.com/dotnet/corefx/blob/a96e44ba9ee62f88a6a6f19cb9fffce24b4eb8ec/eng/configurations/osgroups.props#L37-L41

**Comment by ericstj** [comment]
Somehow I missed that change, yes deriving from linux should be sufficient.

**Comment by safern** [comment]
Addressed feedback. PTAL whenever you guys have some free time.

File: eng/pipelines/linux.yml [hunk 5]
```diff
@@ -134,35 +134,60 @@ stages:
...
-
-              arm64_Debug:
-                _BuildConfig: Debug
-                _architecture: arm64
-                _framework: netcoreapp
-                _dockerContainer: ubuntu_1604_arm64_cross_container
-                _buildScriptPrefix: 'ROOTFS_DIR=/crossrootfs/arm64 '
-                _buildExtraArguments: --warnAsError false
+              # Run these legs on PRs and official builds, but not on CI builds
+              ${{ if or(eq(parameters.fullMatrix, 'false'), eq(parameters.isOfficialBuild, 'true')) }}:
```
Comment (ViktorHofer) [inline eng/pipelines/linux.yml]: This doesn't make sense. We don't run tests on official builds.

Comment (ViktorHofer) [reply eng/pipelines/linux.yml]: OK I was confused becaues you are adding a helixQueue below.

Comment (ViktorHofer) [reply eng/pipelines/linux.yml]: Do we really want to run this in PRs? Why don't we want to run this as part of rolling builds?

Comment (ViktorHofer) [reply eng/pipelines/linux.yml]: Rolling builds always need to be a super-set of PR builds. That's how we measure the CI success rate of PR builds in the CI council.

Comment (ViktorHofer) [reply eng/pipelines/linux.yml]: /end

Comment (safern) [reply eng/pipelines/linux.yml]: > OK I was confused becaues you are adding a helixQueue below.

Yes that was a mistake from copy and paste 😄 -- I will remove it.

> Do we really want to run this in PRs? Why don't we want to run this as part of rolling builds?

Yes, we want this build to be guarded as part of PR builds as the native pieces are different and use a special toolset, we want to catch any issues early. 

I thought of not running them on rolling builds cause it didn't run any tests (yet) and we seem to run tests for all legs in rolling builds, if we're OK with having this leg be part of rolling builds I don't object and can fix that.

Comment (ViktorHofer) [reply eng/pipelines/linux.yml]: Yes please add it to rolling builds as well.

Comment (safern) [reply eng/pipelines/linux.yml]: Done.

**Comment by jkotas** [comment]
> I'm changing it to derive from Linux and set TargetsLinux=true on this PR. 

Is this a temporary hack or is this a permanent logically correct solution?

In other words, are the Linux-specific implementations applicable to WebAssembly? E.g. Is https://github.com/dotnet/corefx/blob/master/src/System.Net.NetworkInformation/src/System/Net/NetworkInformation/LinuxIPGlobalProperties.cs going to work in WebAssembly?

**Comment by safern** [comment]
> Is this a temporary hack or is this a permanent logically correct solution?

I think this would be a long term solution and then if we have a project where we cross-target for Linux and OSX, where we don't have a `Unix` specific implementation, but we need a WebAssembly specific implementation, we would need to add a `netcoreapp-WebAssembly` and then include the specific implementation when `TargetsWebAssembly` == `true`.

However, for the example that you just shared it seems like when we had the support a while ago, it was the same, as `WebAssembly` was importing `Unix` os group, which declared `TargetsLinux=true`. In this project, it is cross compiled for `Linux`, `OSX` and `FreeBSD`, and we didn't have a `TargetsUnix` == `true` specific implementation, so for `WebAssembly` it would've fallback to the `Linux` implementation.

https://github.com/dotnet/corefx/blob/96fccd3abda252d781a447fe801c8b389fa0055f/src/System.Net.NetworkInformation/src/System.Net.NetworkInformation.csproj#L266

I guess as we start running tests and using corefx's implementation fully for WebAssembly we can start applying this fixes and cross-compile for `WebAssembly` if needed, but to me it makes the most sense for `WebAssembly` to be at the same level of `Linux`, which at the end is a subset of `Unix`.

**Comment by jkotas** [comment]
> to me it makes the most sense for WebAssembly to be at the same level of Linux, which at the end is a subset of Unix.

Agree. So are you going to fix it in this PR (ie change WebAssembly to derive from Unix directly)?

**Comment by safern** [comment]
> Agree. So are you going to fix it in this PR (ie change WebAssembly to derive from Unix directly)?

Not right now. Right now the most important piece is to have a starting transport package that mono can consume to use the System.Native pal. Then I think we should start working on using the corefx implementation for `WebAssembly` and try to start running tests for `WebAssembly` as part of that effort as we bring the new platform, I think we should fix-up the implementations and make `WebAssembly` derive from `Unix` directly.

**Comment by jkotas** [comment]
If fixing this properly is too much work, it would be nice to at least add a TODO comment to make it clear that it is not right.

**Review by ericstj — APPROVED** [review]

**Pull request merged**